### DISCUSSION TITLE
[SOLVED]CGLM doesn't compute values as expected

### DISCUSSION THREAD

**Discussion by templalizer1284-zz** [recp/cglm#175] [General]
Hello,

Matrix values are not the same when i am passing data from struct to glm functions compared to by passing raw data.

```
typedef struct Transformation Transformation;

struct Transformation {
	vec3 cam_pos, cam_look, cam_head, axis, scale, translate;
	float fov, aspect_ratio, near_plane, far_plane, angle;
	mat4 Projection, View, Model, MVP;
};

#define TransformationCalculateLocal(x) \
	glm_perspective(x.fov, x.aspect_ratio, x.near_plane, x.far_plane, x.Projection); \
	glm_lookat(x.cam_pos, x.cam_look, x.cam_head, x.View);			\
	glm_mat4_identity(x.Model); \
	glm_scale(x.Model, x.scale); \
	glm_rotate(x.Model, x.angle, x.axis); \
	glm_translate(x.Model, x.translate); \
	glm_mat4_mulN( (mat4 *[]) {&x.Projection, &x.View, &x.Model}, 3, x.MVP );

/* ... later on in main */

	Transformation t_cube = {
		.cam_pos = {4,3,3},
		.cam_look = {0,0,0},
		.cam_head = {0,1,0},
		.axis = {0,0,0},
		.scale = {0.5,0.5,0.5},
		.translate = {1,0,0},
		.fov = glm_rad(45),
		.aspect_ratio = (float)WIDTH / (float)HEIGHT,
		.near_plane = 0.1f,
		.far_plane = 100.0f,
		.angle = 50,
	};

	TransformationCalculateLocal(t_cube);

```

But this works: [Edited, Reason: corrected because i was wrote from another machine]
```
	mat4 Projection, View, Model, MVP;

	glm_perspective(glm_rad(45), (float)1024 / (float)768, 0.1f, 100.0f, Projection);
	glm_lookat( (vec3) {4,3,3}, (vec3) {0,0,0}, (vec3) {0,1,0}, View );
	glm_mat4_identity(Model);
	glm_scale(Model, (vec3) {0.5, 0.5, 0.5});
	glm_rotate(Model, glm_rad(50), (vec3) {0,0,0});
	glm_translate(Model, (vec3) {1,0,0});
	glm_mat4_mulN( (mat4 *[]) {&Projection, &View, &Model}, 3, MVP );
```
When i compare values from t_cube.MVP and MVP they are NOT the same. Where i am making a mistake?

Thank you for your time and your work of making glm possible in C.

-Aleksandar.

**Comment by recp** [comment]
Hi @templalizer1284 

Thank you for starting cglm's first discussion :) 

 > 	glm_mat4_identity(x.Model); \
	glm_scale(x.Model, x.scale); \

You can use `glm_scale_make(x.Model, x.scale)` istead, it creates new scale matrix for you, `glm_scale()` transforms existing one.

**Reply by templalizer1284-zz** [reply]
What do you recommend? is ```glm_scale_make()``` more (general purpose / reliable) for usage?

**Reply by recp** [reply]
If you want to create a scale matrix then `glm_scale_make()` is better to use (no need to set identity matrix first here), if you wan to scale existig matrix/transform then `glm_scale()` is better one. If you want to scale existing matrix with uniform/single value (X == Y == Z) then `glm_scale_uni()` is your friend... hope it helps.

**Comment by recp** [comment]
I'll try to debug your codes

**Comment by recp** [comment]
You used passed angle as `.angle = 50`, `glm_rotate(x.Model, x.angle, x.axis);`. **glm** uses radians not degress, you must convert it to radians as you did below `glm_rad(50)`.  glm_rad(45) != 50.

**Reply by templalizer1284-zz** [reply]
Thank you very much for reply! 

You are right, i didn't put ```glm_rad(50)``` in x.angle... Everything works now. I am very sorry for wasting your time..
I will contribute to your project in any way possible, thank you again for making cglm happen.
